presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/StateCompiler.java

```diff
@@ -178,6 +181,30 @@ public static <T> AccumulatorStateSerializer<T> generateStateSerializer(Class<T>
         }
     }
 
+    private static <T> T getAccumulatorStateMetadataInstance(Class<?> clazz, Map<String, Type> fieldTypes)
+            throws InvocationTargetException, InstantiationException, IllegalAccessException
+    {
+        Optional<Constructor<?>> constructor = Arrays.stream(clazz.getConstructors())
+                // only check for the public constructors
+                .filter(cons -> Modifier.isPublic(cons.getModifiers()))
+                // constructor can only accept Type or nothing
+                .filter(cons -> Arrays.stream(cons.getParameterTypes()).allMatch(Type.class::equals))
+                .filter(cons -> Arrays.stream(cons.getParameters()).allMatch(param -> param.isAnnotationPresent(TypeParameter.class)))
+                .filter(cons -> Arrays.stream(cons.getParameters())
+                        .map(param -> param.getAnnotation(TypeParameter.class).value())
+                        .allMatch(fieldTypes::containsKey))
+                .findFirst();
```

Comment:
Should we validate that one and only one constructor is allowed?

Reply:
I considered this option but chose not to implement it. However, I think it makes sense to reduce ambiguity. I'll add it.

Reply:
Agreed--better for the function author to find out quickly what is supported.

Reply:
The latest commit adds the check as well as a unit test case